### PR TITLE
Add reference for box.stat.memtx().tx

### PR DESCRIPTION
**Pull request by lenkis**

### CODE REVIEW

File: doc/reference/reference_lua/box_stat/memtx.rst [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+.. module:: box.stat
+
+.. function:: memtx()
+
+    Shows memtx-storage-engine activity.
+
+.. _box_introspection-box_stat_memtx_tx:
+
+box.stat.memtx().tx
```
Comment (drewdzzz) [inline doc/reference/reference_lua/box_stat/memtx.rst:14]: Thanks for the doc!

Unfortunately, it seems difficult to read. I think we should avoid full paths (e.g. replace `box.stat.memtx().tx.txn.statements` with `statements` when possible) to reduce documentation size. Also, I would mention `max/avg/total` block as rare as possible.

<details>

<summary> My suggestion </summary>

Note that I removed all the links here so that I could impose it locally. Also, indentation should be probably fixed.


```
box.stat.memtx()
================

    Shows memtx-storage-engine activity.

box.stat.memtx().tx
-------------------

* ``tx`` shows the statistics of the memtx transactional manager,
  which is responsible for transactions (``box.stat.memtx().tx.txn``)
  and multiversion concurrency control (``box.stat.memtx().tx.mvcc``).

  * ``box.stat.memtx().tx.txn`` shows memory allocation related to transactions. Consists of several sections:

    * ``statements`` are *transaction statements*.
      As an example, consider a user starting a transaction with
      ``space:replace{0, 1}`` within this transaction. Under the hood,
      this operation becomes a statement for this transaction.
    * ``user`` is the memory that a user allocated within
      the current transaction using the Tarantool C API function `box_txn_alloc()`.
    * ``system`` is the memory allocated for internal needs (for example, logs) and savepoints.

  * Each sections consists of three parts:

    * ``total`` is the number of bytes that are currently allocated for all transactions in memtx
      (within the section scope).
    * ``avg`` is the average number of bytes that a single transaction uses (equals
      `total`` / number of open transactions).
    * ``max`` is the maximal number of bytes that a single transaction uses.

  * ``box.stat.memtx().tx.mvcc`` shows memory allocation related to multiversion
    concurrency control (MVCC). MVCC is reposible for isolating transactions.
    It reveals conflicts and makes sure that tuples that do not belong to a particular
    space but were (or could be) read by some transaction were not deleted. It consists of three parts:

    * ``trackers`` is the memory allocated for *trackers* of transaction reads. Like in the previous sections,       Tarantool reports the maximal, average and total number of allocated bytes.

    * ``conflicts`` is the memory allocated for *conflicts* which are entities created when transactional conflicts occur. Like in the previous sections, Tarantool reports the maximal, average and total number of allocated bytes.

    * ``tuples`` is the memory allocated for storing tuples.
      With MVCC, tuples are stored using the *stories* mechanism. Nearly every
      tuple has its story. Even tuples in an index may have their stories, so
      it may be useful to differentiate memory allocated for tuples and memory
      allocated for stories.

      All stored tuples fall into three categories, with memory statistics
      reported for each category:

      * ``tracking`` is for tuples that are not used
        by any transactions directly, but MVCC uses them for tracking transaction reads.
      * ``used`` is for tuples that are used
        by active read-write transactions.
      * ``read_view`` is for tuples that are not used
        by active read-write transactions, but are used by read-only transactions.

        For each of the three categories, Tarantool reports two statistical blocks:

        * ``stories`` is for stories.
        * ``retained`` is for *retained* tuples which do not belong to any index,
          but MVCC doesn't allow to delete them yet.

        For each block, Tarantool reports the following statistics:

        * ``count`` is the number of stories or retained tuples.
        * ``total`` is the number of bytes allocated for stories or retained tuples.

```

</details>

Comment (lenkis) [reply doc/reference/reference_lua/box_stat/memtx.rst:14]: Yes, it's a good idea to keep things simple, although reference tends to be more strict and formal than the rest of the doc. I followed the style of a similar reference page about vinyl to keep the text consistent in style and coverage depth: https://docs.d.tarantool.io/en/doc/gh-2801-memtx-tx/reference/reference_lua/box_stat/vinyl/
But I agree, let's make it simple. Thanks!

**Review by drewdzzz — APPROVED** [review]
Thanks for the fixes, LGTM!

File: doc/reference/reference_lua/box_stat.rst [hunk 2]
```diff
@@ -26,6 +26,9 @@ Below is a list of all ``box.stat`` functions.
         *   - :doc:`./box_stat/net`
             - Show network activity
 
+        *   - :doc:`./box_stat/memtx`
+            -  Show memtx-storage-engine activity
+
         *   - :doc:`./box_stat/vinyl`
             -  Show vinyl-storage-engine activity
```
Comment (p7nov) [inline doc/reference/reference_lua/box_stat.rst]:
```suggestion
            -  Show ``memtx`` storage engine activity

        *   - :doc:`./box_stat/vinyl`
            -  Show ``vinyl`` storage engine activity
```

File: doc/reference/reference_lua/box_stat/memtx.rst [hunk 3]
```diff
@@ -0,0 +1,81 @@
+.. _box_introspection-box_stat_memtx:
+
+box.stat.memtx()
+================
+
+.. module:: box.stat
+
+.. function:: memtx()
+
+    Shows memtx-storage-engine activity.
```
Comment (p7nov) [inline doc/reference/reference_lua/box_stat/memtx.rst]:
```suggestion
    Shows ``memtx`` storage engine activity.
```

File: doc/reference/reference_lua/box_stat/memtx.rst [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+box.stat.memtx().tx
+-------------------
+
+* ``tx`` shows the statistics of the memtx transactional manager,
+  which is responsible for transactions (``box.stat.memtx().tx.txn``)
+  and multiversion concurrency control (``box.stat.memtx().tx.mvcc``).
+
+  * ``box.stat.memtx().tx.txn`` shows memory allocation related to transactions.
+
+    It consists of the following sections:
```
Comment (p7nov) [inline doc/reference/reference_lua/box_stat/memtx.rst]: Not sure if `sections` is the right word here. If it's an object, these may be its _fields_ or something similar.

File: doc/reference/reference_lua/box_stat/memtx.rst [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+    * ``statements`` are *transaction statements*.
+      As an example, consider a user starting a transaction with
+      ``space:replace{0, 1}`` within this transaction. Under the hood,
```
Comment (p7nov) [inline doc/reference/reference_lua/box_stat/memtx.rst]: The example is hard for me to grasp. A user starts a transaction with a `space:replace` call - I get it. What is the meaning of `within this transaction` ending?
Can you maybe rephrase?

File: doc/reference/reference_lua/box_stat/memtx.rst [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    * ``system`` is the memory allocated for internal needs
+      (for example, logs) and savepoints.
+
+      For each section, Tarantool reports the following statistics:
+
+      * ``total`` is the number of bytes that are currently allocated in memtx
+        for all transactions within the section scope.
+      * ``avg`` is the average number of bytes that a single transaction uses
+        (equals ``total`` / number of open transactions).
+      * ``max`` is the maximal number of bytes that a single transaction uses.
```
Comment (p7nov) [inline doc/reference/reference_lua/box_stat/memtx.rst]:
```suggestion
      * ``max`` is the maximum number of bytes that a single transaction uses.
```

Sounds more correct to me. Like in ``maximum speed``, ``maximum value``. I checked the difference here ["Maximum" vs. "maximal"](https://english.stackexchange.com/questions/38260/maximum-vs-maximal).

File: doc/reference/reference_lua/box_stat/memtx.rst [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+  * ``box.stat.memtx().tx.mvcc`` shows memory allocation related to multiversion
+    concurrency control (MVCC). MVCC is reponsible for isolating transactions.
```
Comment (p7nov) [inline doc/reference/reference_lua/box_stat/memtx.rst]: Maybe a link to https://www.tarantool.io/en/doc/latest/platform/atomic/txn_mode_mvcc/ ?

File: doc/reference/reference_lua/box_stat/memtx.rst [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    It reveals conflicts and makes sure that tuples that do not belong to a particular
+    space but were (or could be) read by some transaction were not deleted.
+
+    It consists of the following sections:
+
+    * ``trackers`` is the memory allocated for *trackers* of transaction reads.
+      Like in the previous sections, Tarantool reports the total, average
```
Comment (p7nov) [inline doc/reference/reference_lua/box_stat/memtx.rst]: Nit: The introductory clause `Like in the previous sections` sounds a bit awkward to me. As a reference reader, I would just search (ctrl+F) for a specific name, not read the whole page. Thus, `previous sections` without a clickable link forces me to start reading the content I didn't need)

Maybe, create a subsection with `total`, `avg`, and `max` definitions and link it from all applicable objects?

File: doc/reference/reference_lua/box_stat/memtx.rst [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        For each of the three categories, Tarantool reports two statistical blocks:
+
+        * ``stories`` is for stories.
+        * ``retained`` is for *retained* tuples which do not belong to any index,
+          but MVCC doesn't allow to delete them yet.
+
+        For each block, Tarantool reports the following statistics:
+
+        * ``count`` is the number of stories or retained tuples.
+        * ``total`` is the number of bytes allocated for stories or retained tuples.
```
Comment (p7nov) [inline doc/reference/reference_lua/box_stat/memtx.rst]: The described object and field hierarchy is rather deep for a text explanation. Can we just show two output examples (for txn and mvcc)?

**Review by p7nov — COMMENTED** [review]
Some improvement suggestions and ideas from my side

**Comment by p7nov** [comment]
There is also [box.stat.reset()](https://docs.d.tarantool.io/en/doc/gh-2801-memtx-tx/reference/reference_lua/box_stat/reset/) function. Does it reset new memtx stats?

File: doc/reference/reference_lua/box_stat/memtx.rst (hunk 4 above)
Comment (lenkis) [reply doc/reference/reference_lua/box_stat/memtx.rst]: I'd keep "sections" because we refer to them below. "Field" sounds awkward in this context, for example:

```
      For each section, Tarantool reports the following statistics:

      * ``total`` is the number of bytes that are currently allocated in memtx
        for all transactions within the section scope.
```

File: doc/reference/reference_lua/box_stat/memtx.rst (hunk 5 above)
Comment (lenkis) [reply doc/reference/reference_lua/box_stat/memtx.rst]: Let me try :) is it ok?

```
      As an example, consider a user starting a transaction with
      ``space:replace{0, 1}`` being one of the instructions within
      this transaction.
```

File: doc/reference/reference_lua/box_stat/memtx.rst (hunk 9 above)
Comment (lenkis) [reply doc/reference/reference_lua/box_stat/memtx.rst]: Good point to add examples, but I would keep the text as well.

File: doc/reference/reference_lua/box_stat/memtx.rst [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+
+**Example**
+
+Let's get memory statistics for `used` tuples in a transaction.
+
+To make transactions work, we need to :ref:`enable the MVCC engine <txn_mode_mvcc-enabling>`.
```
Comment (drewdzzz) [inline doc/reference/reference_lua/box_stat/memtx.rst]: > To make transactions work

Actually, transactions always work, so let's rephrase it. Probably:

Let's enable MVCC so that `box.stat.memtx.tx().mvcc` will contain non-zero values.

File: doc/reference/reference_lua/box_stat/memtx.rst (hunk 6 above)
Comment (p7nov) [reply doc/reference/reference_lua/box_stat/memtx.rst]: I see the reaction (👍 ) but not the change. Do you agree?)
image

File: doc/reference/reference_lua/box_stat/memtx.rst [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+      For each block, Tarantool reports the following statistics:
+
+      * ``count`` is the number of stories or retained tuples.
+      * ``total`` is the number of bytes allocated for stories or retained tuples.
+
+.. _box_introspection-box_stat_memtx_tx_example:
+
+**Example**
+
+Let's get memory statistics for ``used`` tuples in a transaction.
```
Comment (p7nov) [inline doc/reference/reference_lua/box_stat/memtx.rst]: In the modern [Tarantool docs style](https://www.tarantool.io/en/doc/latest/contributing/docs/style/) we don't use first person: `Let's`, `we`, and so on. Better to rephrase this.

File: doc/reference/reference_lua/box_stat/memtx.rst [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+
+First, we :ref:`enable MVCC <txn_mode_mvcc-enabling>` so that
+``box.stat.memtx.tx().mvcc`` contained non-zero values.
+
+.. code-block:: lua
+
+   box.cfg{memtx_use_mvcc_engine = true}
+
+We did it within the first ``box.cfg{}`` call to a new Tarantool instance,
```
Comment (p7nov) [inline doc/reference/reference_lua/box_stat/memtx.rst]: `box.cfg` is considered legacy since 3.0. It also may be dangerous when mixed with YAML config. We avoid creating new content with this approach. I'd just say that the cluster must be started with [database.use_mvcc_engine](https://www.tarantool.io/en/doc/latest/reference/configuration/configuration_reference/#confval-database.use_mvcc_engine) set to `true`.

**Review by p7nov — CHANGES_REQUESTED** [review]

File: doc/reference/reference_lua/box_stat/memtx.rst [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+
+.. _box_introspection-box_stat_memtx_tx_example:
+
+**Example**
+
+This example illustrates memory statistics for ``used`` tuples in a transaction.
+
+The cluster must be started with the :ref:`database.use_mvcc_engine <configuration_reference_database_use_mvcc_engine>`
+parameter set to true. This :ref:`enables MVCC <txn_mode_mvcc-enabling>` so that
+``box.stat.memtx.tx().mvcc`` contained non-zero values.
```
Comment (p7nov) [inline doc/reference/reference_lua/box_stat/memtx.rst]:
```suggestion
``box.stat.memtx.tx().mvcc`` contains non-zero values.
```

File: doc/reference/reference_lua/box_stat/memtx.rst [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+
+In the transaction above, three tuples are replaced by the `0` key:
+
+* ``{0, 0}``
+* ``{0, 'aa...aa'}``
+* ``{0, 1}``
+
+MVCC considers all these tuples as ``used`` since they belong to the current transaction.
+Also, MVCC considers tuples ``{0, 0}`` and ``{0, 'aa..aa'}`` as ``retained`` because
+they don't belong to any index (unlike ``{0, 1}``), but they cannot be deleted yet.
```
Comment (p7nov) [inline doc/reference/reference_lua/box_stat/memtx.rst]:
```suggestion
they don't belong to any index (unlike ``{0, 1}``) but cannot be deleted yet.
```

File: doc/reference/reference_lua/box_stat/memtx.rst [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+
+Calling ``box.stat.memtx.tx()`` now will bring something like this:
```
Comment (p7nov) [inline doc/reference/reference_lua/box_stat/memtx.rst]:
```suggestion
Calling ``box.stat.memtx.tx()`` now returns the following result:
```

Up to you. As I know, "something like this" is often considered weak or vague writing in tech docs.

**Review by p7nov — APPROVED** [review]
LGTM, just a couple fixes in suggestions.
